In [118]:
import pandas as pd
import warnings 
warnings.filterwarnings('ignore')


In [119]:
# load the dataset
df = pd.read_csv('HRDataset_v14.csv')
pd.options.display.max_columns = None
df.head()

Employee_Name  EmpID  MarriedID  MaritalStatusID  GenderID  \
0       Adinolfi, Wilson  K  10026          0                0         1   
1  Ait Sidi, Karthikeyan     10084          1                1         1   
2         Akinkuolie, Sarah  10196          1                1         0   
3              Alagbe,Trina  10088          1                1         0   
4          Anderson, Carol   10069          0                2         0   

   EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  Salary  Termd  \
0            1       5            4                       0   62506      0   
1            5       3            3                       0  104437      1   
2            5       5            3                       0   64955      1   
3            1       5            3                       0   64991      0   
4            5       5            3                       0   50825      1   

   PositionID                  Position State   Zip        DOB Sex  \
0          19   Production Technician I    MA  1960  7/10/1983  M    
1          27                   Sr. DBA    MA  2148   5/5/1975  M    
2          20  Production Technician II    MA  1810  9/19/1988   F   
3          19   Production Technician I    MA  1886  9/27/1988   F   
4          19   Production Technician I    MA  2169   9/8/1989   F   

  MaritalDesc CitizenDesc HispanicLatino RaceDesc DateofHire  \
0      Single  US Citizen             No    White   7/5/2011   
1     Married  US Citizen             No    White  3/30/2015   
2     Married  US Citizen             No    White   7/5/2011   
3     Married  US Citizen             No    White   1/7/2008   
4    Divorced  US Citizen             No    White  7/11/2011   

  DateofTermination         TermReason        EmploymentStatus  \
0               NaN  N/A-StillEmployed                  Active   
1         6/16/2016      career change  Voluntarily Terminated   
2         9/24/2012              hours  Voluntarily Terminated   
3               NaN  N/A-StillEmployed                  Active   
4          9/6/2016   return to school  Voluntarily Terminated   

          Department     ManagerName  ManagerID RecruitmentSource  \
0  Production         Michael Albert       22.0          LinkedIn   
1              IT/IS      Simon Roup        4.0            Indeed   
2  Production         Kissy Sullivan       20.0          LinkedIn   
3  Production           Elijiah Gray       16.0            Indeed   
4  Production         Webster Butler       39.0     Google Search   

  PerformanceScore  EngagementSurvey  EmpSatisfaction  SpecialProjectsCount  \
0          Exceeds              4.60                5                     0   
1      Fully Meets              4.96                3                     6   
2      Fully Meets              3.02                3                     0   
3      Fully Meets              4.84                5                     0   
4      Fully Meets              5.00                4                     0   

  LastPerformanceReview_Date  DaysLateLast30  Absences  
0                  1/17/2019               0         1  
1                  2/24/2016               0        17  
2                  5/15/2012               0         3  
3                   1/3/2019               0        15  
4                   2/1/2016               0         2

In [120]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311 entries, 0 to 310
Data columns (total 36 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Employee_Name               311 non-null    object 
 1   EmpID                       311 non-null    int64  
 2   MarriedID                   311 non-null    int64  
 3   MaritalStatusID             311 non-null    int64  
 4   GenderID                    311 non-null    int64  
 5   EmpStatusID                 311 non-null    int64  
 6   DeptID                      311 non-null    int64  
 7   PerfScoreID                 311 non-null    int64  
 8   FromDiversityJobFairID      311 non-null    int64  
 9   Salary                      311 non-null    int64  
 10  Termd                       311 non-null    int64  
 11  PositionID                  311 non-null    int64  
 12  Position                    311 non-null    object 
 13  State                       311 non

In [121]:
df.isnull().sum()

Employee_Name                   0
EmpID                           0
MarriedID                       0
MaritalStatusID                 0
GenderID                        0
EmpStatusID                     0
DeptID                          0
PerfScoreID                     0
FromDiversityJobFairID          0
Salary                          0
Termd                           0
PositionID                      0
Position                        0
State                           0
Zip                             0
DOB                             0
Sex                             0
MaritalDesc                     0
CitizenDesc                     0
HispanicLatino                  0
RaceDesc                        0
DateofHire                      0
DateofTermination             207
TermReason                      0
EmploymentStatus                0
Department                      0
ManagerName                     0
ManagerID                       8
RecruitmentSource               0
PerformanceSco

* DateofTermination has 207 missing values. This could suggest that some employees are still active or the termination dates are missing for certain records.
* ManagerID has 8 missing values, indicating that some employees do not have an assigned manager or this data is incomplete for a few records.

In [122]:
df[df.ManagerID.isnull()][['ManagerID', "ManagerName", 'Department']]

ManagerID     ManagerName         Department
19         NaN  Webster Butler  Production       
30         NaN  Webster Butler  Production       
44         NaN  Webster Butler  Production       
88         NaN  Webster Butler  Production       
135        NaN  Webster Butler  Production       
177        NaN  Webster Butler  Production       
232        NaN  Webster Butler  Production       
251        NaN  Webster Butler  Production

* The ManagerName for all the rows with missing ManagerID is "Webster Butler". This suggests that the ManagerID for these employees may not have been recorded in the dataset, or perhaps there's a mismatch or inconsistency in how ManagerIDs are assigned for employees in the Production department.

In [123]:
df[df.ManagerName== 'Webster Butler'][['ManagerID', "ManagerName", 'Department']]

ManagerID     ManagerName         Department
4         39.0  Webster Butler  Production       
19         NaN  Webster Butler  Production       
30         NaN  Webster Butler  Production       
44         NaN  Webster Butler  Production       
65        39.0  Webster Butler  Production       
88         NaN  Webster Butler  Production       
89        39.0  Webster Butler  Production       
105       39.0  Webster Butler  Production       
124       39.0  Webster Butler  Production       
135        NaN  Webster Butler  Production       
151       39.0  Webster Butler  Production       
174       39.0  Webster Butler  Production       
177        NaN  Webster Butler  Production       
198       39.0  Webster Butler  Production       
206       39.0  Webster Butler  Production       
214       39.0  Webster Butler  Production       
232        NaN  Webster Butler  Production       
251        NaN  Webster Butler  Production       
276       39.0  Webster Butler  Production       
280       39.0  Webster Butler  Production       
300       39.0  Webster Butler  Production

* It appears that Webster Butler has a ManagerID of 39.0 in some records, but in others, the ManagerID is missing (NaN). This inconsistency may be due to incomplete or incorrect data entry for certain employees in the Production department.

In [124]:
df['ManagerID'].fillna(39, inplace = True)

* Since Webster Butler has a ManagerID of 39, I have filled the missing ManagerID values (for rows where the manager is Webster Butler) with 39.0. This would ensure consistency and correct assignment of the manager to these employees.

In [125]:
df.ManagerID = df.ManagerID.astype('int') 

In [126]:

date_columns = ['DOB', 'DateofHire', 'DateofTermination', 'LastPerformanceReview_Date']
df[date_columns].dtypes

DOB                           object
DateofHire                    object
DateofTermination             object
LastPerformanceReview_Date    object
dtype: object

In [127]:
# I changed the date colunms data type to datecolumns
for i in date_columns:
    df[i] = df[i].astype('datetime64[ns]')

In [128]:
# Rearrange and format
df['Employee_Name'] = (df['Employee_Name'].str.split(',', expand = True).iloc[:,1] 
                    + ' '
                    + df['Employee_Name'].str.split(',', expand = True).iloc[:,0])

* Rearrange and format the Employee_Name column, where the names were written as Last Name, First Name, with a comma separating the last name and first name to First Name Last Name

In [129]:
# Dropped redundant columns
df.drop(columns = ['MarriedID','MaritalStatusID', 'GenderID', 
                   'Termd', 'PerfScoreID', 'EmpStatusID',
                   'PositionID',], inplace = True)


In [130]:
# reorder of columns
cols_reorder = ['EmpID', 'Employee_Name','Sex', 'MaritalDesc', 'DeptID',
                'FromDiversityJobFairID', 'Salary',
                'Position', 'State', 'Zip', 'DOB', 
                'CitizenDesc', 'HispanicLatino', 'RaceDesc', 'DateofHire',
                'DateofTermination', 'TermReason', 'EmploymentStatus', 'Department',
                'ManagerName', 'ManagerID', 'RecruitmentSource', 'PerformanceScore',
                'EngagementSurvey', 'EmpSatisfaction', 'SpecialProjectsCount',
                'LastPerformanceReview_Date', 'DaysLateLast30', 'Absences'
       ]

In [131]:
df = df[cols_reorder]
df.head()

EmpID             Employee_Name Sex MaritalDesc  DeptID  \
0  10026        Wilson  K Adinolfi  M       Single       5   
1  10084   Karthikeyan    Ait Sidi  M      Married       3   
2  10196          Sarah Akinkuolie   F     Married       5   
3  10088              Trina Alagbe   F     Married       5   
4  10069           Carol  Anderson   F    Divorced       5   

   FromDiversityJobFairID  Salary                  Position State   Zip  \
0                       0   62506   Production Technician I    MA  1960   
1                       0  104437                   Sr. DBA    MA  2148   
2                       0   64955  Production Technician II    MA  1810   
3                       0   64991   Production Technician I    MA  1886   
4                       0   50825   Production Technician I    MA  2169   

         DOB CitizenDesc HispanicLatino RaceDesc DateofHire DateofTermination  \
0 1983-07-10  US Citizen             No    White 2011-07-05               NaT   
1 1975-05-05  US Citizen             No    White 2015-03-30        2016-06-16   
2 1988-09-19  US Citizen             No    White 2011-07-05        2012-09-24   
3 1988-09-27  US Citizen             No    White 2008-01-07               NaT   
4 1989-09-08  US Citizen             No    White 2011-07-11        2016-09-06   

          TermReason        EmploymentStatus         Department  \
0  N/A-StillEmployed                  Active  Production          
1      career change  Voluntarily Terminated              IT/IS   
2              hours  Voluntarily Terminated  Production          
3  N/A-StillEmployed                  Active  Production          
4   return to school  Voluntarily Terminated  Production          

      ManagerName  ManagerID RecruitmentSource PerformanceScore  \
0  Michael Albert         22          LinkedIn          Exceeds   
1      Simon Roup          4            Indeed      Fully Meets   
2  Kissy Sullivan         20          LinkedIn      Fully Meets   
3    Elijiah Gray         16            Indeed      Fully Meets   
4  Webster Butler         39     Google Search      Fully Meets   

   EngagementSurvey  EmpSatisfaction  SpecialProjectsCount  \
0              4.60                5                     0   
1              4.96                3                     6   
2              3.02                3                     0   
3              4.84                5                     0   
4              5.00                4                     0   

  LastPerformanceReview_Date  DaysLateLast30  Absences  
0                 2019-01-17               0         1  
1                 2016-02-24               0        17  
2                 2012-05-15               0         3  
3                 2019-01-03               0        15  
4                 2016-02-01               0         2

In [132]:
# saving cleaned data
df.to_csv('HR_Employee_Cleaned_Data.csv', index = False)